third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -236,6 +247,123 @@ impl<'a> FunctionGenerator<'a> {
         }
     }
 
+    /// Remove pairs of instructions that are equivalent to a no-op.
+    fn remove_redundant_instr_pairs(&mut self) {
```

Comment:
Are we planning to move this out of this already pretty large file? I thought we'd implement it as a transformation on file format bytecode. Fine for now, but if the grows, thats not good here. We also like to have a design where its easy to plugin new optimizations. We could come up with something along:

```
trait PeepHoleOptimizer {
  fn optimize(basic_block: Vec<FF::Bytecode>) -> Vec<FF::Bytecode>
}

trait FixedWindowPeepholeOptimizer {
  const WINDOW_SIZE: usize;
  
  fn optimize_window(win: Vec<FF::ByteCode) -> Option<Vec<FF::Bytecode>;
  
  fn optimize(...) { ... }
```

Reply:
Thanks for the abstraction idea.

Done, but slightly differently (because `trait FixedWindowPeepholeOptimizer : PeepHoleOptimizer` is not trait inheritance, but a requirement that both traits should be implemented by anyone implementing the first).

I think it achieves the same goal overall.